src/dataflow/src/source/mod.rs

```diff
@@ -1361,23 +1361,32 @@ where
                     //
                     // Side note: with persistence enabled for the source, we will anyways never
                     // re-emit old data, because we start reading from the persisted offsets.
-                    let current_binding = timestamp_histories
-                        .get_binding(&source_ts.partition, source_ts.offset)
-                        .map(|(binding, _offset)| binding)
-                        .unwrap_or(0);
-                    if current_binding < assigned_ts.0 {
-                        timestamp_histories.add_binding(
-                            source_ts.partition.clone(),
-                            assigned_ts.0,
-                            source_ts.offset,
-                            false,
-                        );
-                    } else {
-                        tracing::debug!(
-                            "Filtered out timestamp binding {:?} from persistence because we already have {}.",
-                            (source_ts, assigned_ts),
-                            current_binding
-                        );
+                    let lower = Antichain::from_elem(assigned_ts.0);
+                    let upper = Antichain::new();
+                    // NOTE: We use get_bindings_in_range() and not the seemingly better
```

Comment:
I don't fully understand what was happening before - as in, this bit in the commit message:

```
This could lead to the case
were we thought that there we already bindings that supersede the
bindings restored from persistence but there were not, and we would
therefore end up in a situation without any bindings, meaning the Kafka
source would re-read all its input from the beginning again.
```

is a lot of moving parts

it seems like

1. if bindings from sqlite supersede bindings from persist we throw away bindings from persist?
2. somehow we only use the bindings from persist to decide what data we already have?
3. based on 2. we use the new proposed bindings and end up re-reading from kafka?

does that sound right? I didn't yet reread the rest of the surrounding code, but I can do that later. If there's any way to add a test for this bug that would be _amazing_. No worries if not though.

Reply:
Should we do a call on Monday? I can walk you through it. I think it's valuable to properly think it through, no need to rush.

Side note: there is already a test, Philips persistence tests that use a Kafka source caught this. But it only surfaced once compaction was enabled, which I can also probably better explain in a call.

Reply:
I'd love that! I don't want to take away from your skunkworks but I can do a call today if you'd rather move faster on this. I'm happy either way totally up to you

Reply:
Let's do Monday :ok_hand: